File: dotnet/core/workspace/csharp/adapters/allors.workspace.adapters.remote/session/originstate/DatabaseOriginState.cs
// <copyright file="DatabaseOriginState.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters.Remote
{
    using System.Collections.Generic;
    using Allors.Protocol.Json.Api.Push;

    internal sealed class DatabaseOriginState : Adapters.DatabaseOriginState
    {
        internal DatabaseOriginState(Strategy strategy, DatabaseRecord record) : base(record) => this.RemoteStrategy = strategy;

        public override Adapters.Strategy Strategy => this.RemoteStrategy;
        private Strategy RemoteStrategy { get; }

        internal PushRequestNewObject PushNew() => new PushRequestNewObject
        {
            w = this.Id,
            t = this.Class.Tag,
            r = this.PushRoles()
        };

        internal PushRequestObject PushExisting() => new PushRequestObject
        {
            d = this.Id,
            v = this.Version,
            r = this.PushRoles()
        };

        private PushRequestRole[] PushRoles()
        {
            if (this.ChangedRoleByRelationType?.Count > 0)
            {
                var database = this.RemoteStrategy.Session.Workspace.DatabaseConnection;

                var numbers = database.Numbers;

                var roles = new List<PushRequestRole>();

                foreach (var keyValuePair in this.ChangedRoleByRelationType)
                {
                    var relationType = keyValuePair.Key;
                    var roleValue = keyValuePair.Value;

                    var pushRequestRole = new PushRequestRole { t = relationType.Tag };

                    if (relationType.RoleType.ObjectType.IsUnit)
                    {
                        pushRequestRole.u = database.UnitConvert.ToJson(roleValue);
                    }
                    else if (relationType.RoleType.IsOne)
                    {
                        pushRequestRole.c = (long?)roleValue;
                    }
                    else
                    {
                        if (!this.ExistDatabaseRecord)
                        {
                            pushRequestRole.a = numbers.ToArray(roleValue);
                        }
                        else
                        {
                            var databaseRole = (long[])this.DatabaseRecord.GetRole(relationType.RoleType);
                            if (databaseRole == null)
                            {
                                pushRequestRole.a = numbers.ToArray(roleValue);
                            }
                            else
                            {
                                pushRequestRole.a = numbers.ToArray(numbers.Except(roleValue, databaseRole));
                                pushRequestRole.r = numbers.ToArray(numbers.Except(databaseRole, roleValue));
                            }
                        }
                    }

                    roles.Add(pushRequestRole);
                }

                return roles.ToArray();
            }

            return null;
        }
    }
}

File: typescript/libs/workspace/adapters/json/system/src/lib/session/originstate/DatabaseOriginState.ts
import { PushRequestNewObject, PushRequestObject, PushRequestRole } from '@allors/protocol/json/system';
import { DatabaseOriginState as SystemDatabaseOriginState, DatabaseRecord, difference, Strategy, save, IRange, importFrom } from '@allors/workspace/adapters/system';
import { unitToJson } from '../../json/toJson';

export class DatabaseOriginState extends SystemDatabaseOriginState {
  constructor(public strategy: Strategy, record: DatabaseRecord) {
    super(record);
  }

  pushNew(): PushRequestNewObject {
    return {
      w: this.id,
      t: this.class.tag,
      r: this.pushRoles(),
    };
  }

  pushExisting(): PushRequestObject {
    return {
      d: this.id,
      v: this.version,
      r: this.pushRoles(),
    };
  }

  private pushRoles(): PushRequestRole[] {
    if (this.xchangedRoleByRelationType?.size > 0) {
      const roles: PushRequestRole[] = [];

      for (const [relationType, roleValue] of this.xchangedRoleByRelationType) {
        const pushRequestRole: PushRequestRole = { t: relationType.tag };

        if (relationType.roleType.objectType.isUnit) {
          pushRequestRole.u = unitToJson(roleValue);
        } else if (relationType.roleType.isOne) {
          pushRequestRole.c = (roleValue as Strategy)?.id;
        } else {
          const roleIds = importFrom([...(roleValue as Set<Strategy>)].map((v) => v.id));
          if (!this.existRecord) {
            pushRequestRole.a = save(roleIds);
          } else {
            const databaseRole = this.databaseRecord.getRole(relationType.roleType) as IRange;
            if (databaseRole == null) {
              pushRequestRole.a = save(roleIds);
            } else {
              pushRequestRole.a = save(difference(roleIds, databaseRole));
              pushRequestRole.r = save(difference(databaseRole, roleIds));
            }
          }
        }

        roles.push(pushRequestRole);
      }

      return roles;
    }

    return null;
  }
}
